### PR TITLE
[Fix] Outcomes IndexedDB schema

### PR DESCRIPTION
**Pull request by jkasten2**
# Description
## 1 Line Summary
Fix errors saving clicked outcomes due to keyPath value changed incorrectly.

## Details
We are doing the following indexedDB table renames in this commit:
* NotificationClicked -> Outcomes.NotificationClicked
   * Correcting keyPath to "notificationId" in the new table.
* NotificationReceived -> Outcomes.NotificationReceived

## Modiviation
Fixes the notification click direct outcomes feature. Also it is unknown if the indexedDB errors are causing throwing that is resulting in any other side-effect issues due to errors not being caught.

This is done to correct the keyPath for the "NotificationClicked" indexedDB table, you can't change it so a new table must be created.

## Background
Table was created with wrong keyPath of "notification.id" for new visitors for versions from 160000.beta4 to 160000. Writes were attempted as "notificationId" in released 160000 however they may have failed if the visitor was new when those releases were in the wild. However those new on 160000.beta4 to 160000.beta8 will have records saved as "notification.id" that will be converted here.

See related tickets below for more details on the history of the issue.

# Validation
### Checklist
   - [x] All the automated tests pass or I explained why that is not possible
   - [X] I have personally tested this on my machine or explained why that is not possible
   - [x] I have included test coverage for these changes or explained why they are not needed

**Programming Checklist**
Interfaces:
   - [X] Don't use default export
   - [X] New interfaces are in model files

Functions:
   - [X] Don't use default export
   - [X] All function signatures have return types
   - [X] Helpers should not access any data but rather be given the data to operate on.

Typescript:
   - [X] No Typescript warnings
   - [X] Avoid silencing null/undefined warnings with the exclamation point

Other:
   - [X] Iteration: refrain from using `elem of array` syntax. Prefer `forEach` or use `map`
   - [X] Avoid using global OneSignal accessor for `context` if possible. Instead, we can pass it to function/constructor so that we don't call `OneSignal.context`

### Checklist
   - [X] I have included screenshots/recordings of the intended results or explained why they are not needed

---

## Related Tickets
* PR #1028 - Introduced a keyPath from "notificationId" -> "notification.id" without a migration which created this issue:
   -  https://github.com/OneSignal/OneSignal-Website-SDK/pull/1028/files#diff-dd2c30ab15a2323012089532df9ab057c1861af04fa6022f1af434781885d39bL114
* #1078 - Changed back to "notificationId" for reads / writes but also did not address the keyPath schema issue.
---

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Website-SDK/1087)

### CODE REVIEW

**Review by rgomezp — CHANGES_REQUESTED** [review]

File: src/shared/services/Database.ts [hunk 1]
```diff
@@ -40,12 +40,13 @@ interface DatabaseResult {
...
- * "NotificationClicked" = Outcomes only, notifications part of it's session
- * "NotificationReceived" = Outcomes only, notifications part of it's session
  */
 
+export const TABLE_OUTCOMES_NOTIFICATION_CLICKED = "NotificationClicked";
+export const TABLE_OUTCOMES_NOTIFICATION_RECEIVED = "NotificationReceived";
+
 export type OneSignalDbTable = "Options" | "Ids" | "NotificationOpened" | "Sessions" |
-  "NotificationOpened" | "NotificationReceived" | "NotificationClicked" | "SentUniqueOutcome" |  ModelName;
+  "NotificationOpened" | typeof TABLE_OUTCOMES_NOTIFICATION_RECEIVED | typeof TABLE_OUTCOMES_NOTIFICATION_CLICKED | "SentUniqueOutcome" |  ModelName;
```
[the remark this answers is not in the document]
Comment (jkasten2) [reply src/shared/services/Database.ts:49]: `typeof` in Javascript always executes, but in Typescript `typeof` as some cases where [it's compile time](https://www.typescriptlang.org/docs/handbook/2/typeof-types.html).

The use of use of `typeof`  on something like `const MY_CONST` is another one of those compile time cases. See this exact usages here: https://stackoverflow.com/a/56263257

File: src/shared/services/IndexedDb.ts [hunk 2]
```diff
@@ -132,6 +146,74 @@ export default class IndexedDb {
...
+  private migrateOutcomesNotificationReceivedTableForV5(
+    db: IDBDatabase,
+    transaction: IDBTransaction,
+  ) {
+    const newTableName = "Outcomes.NotificationReceived";
+    db.createObjectStore(newTableName, { keyPath: "notificationId" });
+
+    const oldTableName = "NotificationReceived"
+    const cursor = transaction.objectStore(oldTableName).openCursor();
+    cursor.onsuccess = (event: any) => { // Using any here as the TypeScript definition is wrong
```
[the remark this answers is not in the document]
Comment (jkasten2) [reply src/shared/services/IndexedDb.ts:211]: Not exactly the exact same issue, but there is reported issue with the same thing on the `open` `indexedDB` api:
https://github.com/microsoft/TypeScript/issues/28293

It does have a nice work around that seems to also work here. I can update it to use it so we don't have to use any here.

File: src/shared/services/IndexedDb.ts [hunk 3]
```diff
@@ -132,6 +146,74 @@ export default class IndexedDb {
...
+        // Delete old table once we have gone through all records
+        db.deleteObjectStore(oldTableName);
+        return;
+      }
+      transaction
+        .objectStore(newTableName)
+        .put(cursorResult.value);
+      cursorResult.continue();
+    };
+    cursor.onerror = () => { throw cursor.error; };
```
[the remark this answers is not in the document]
Comment (jkasten2) [reply src/shared/services/IndexedDb.ts]: hmm, ya better to just log the error and move on, as throwing can prevent other migrations.

There isn't really anything that can be done now, we can keep the old table around if it fails. That way we have something to look at and debug, as possibly come up with a future fix.

Made these changes in a new commit

File: src/shared/services/IndexedDb.ts [hunk 4]
```diff
@@ -115,6 +120,7 @@ export default class IndexedDb {
       // "{ keyPath: "notification.id" }". This resulted in DB v4 either
       // having "notificationId" or "notification.id" depending if the visitor
       // was new while this version was live.
+      // DB v5 was create to trigger a migration to fix this bug.
```
[the remark this answers is not in the document]
Comment (jkasten2) [reply src/shared/services/IndexedDb.ts]: fixed

File: __test__/unit/services/indexedDb.test.ts [hunk 5]
```diff
@@ -1,23 +1,105 @@
...
+      const openDbRequest = indexedDB.open(dbName, 4);
+      const dbOpenPromise = new Promise((resolve) => {
+        openDbRequest.onsuccess = resolve;
+      });
+      const dbUpgradePromise = new Promise<void>((resolve) => {
+        openDbRequest.onupgradeneeded = function (event: any) {
+          const target = event.target as IDBOpenDBRequest;
+          const db = target.result;
+          db.createObjectStore("NotificationClicked", { keyPath: "notification.id" });
+          db.createObjectStore("NotificationReceived", { keyPath: "notificationId" });
```
Comment (rgomezp) [inline __test__/unit/services/indexedDb.test.ts:84]: what's the point of this line if we're not testing the migration for "NotificationReceived" below?

File: src/shared/services/IndexedDb.ts [hunk 6]
```diff
@@ -4,15 +4,15 @@ import Utils from '../context/Utils';
 import Emitter from '../libraries/Emitter';
 import Log from '../libraries/Log';
 
-const DATABASE_VERSION = 5;
```
Comment (rgomezp) [inline src/shared/services/IndexedDb.ts]: maybe we should hold onto this and simply set it as the default in the constructor. but this pattern of setting the version of a constant will make it easier to find & update in the future

File: __test__/unit/services/indexedDb.test.ts (hunk 5 above)
Comment (jkasten2) [reply __test__/unit/services/indexedDb.test.ts:84]: Test is setting up the DB for how it should look for the v4 to v5 migration in production, even if we are not testing the NotificationReceived in this specific test. The test doesn't work without this line either, as the onupgradeneeded won't finish without as `migrateOutcomesNotificationReceivedTableForV5` throws.

File: src/shared/services/IndexedDb.ts (hunk 6 above)
Comment (jkasten2) [reply src/shared/services/IndexedDb.ts]: ya keeping that as a constant make finding it easier. I added a `--fixup` commit to address it.

**Review by rgomezp — APPROVED** [review]

**Pull request merged**